src/zkAppWorker.ts

```diff
@@ -0,0 +1,162 @@
+import {
+  CircuitString,
+  type Field,
+  MerkleTree,
+  Poseidon,
+  type Proof,
+} from "snarkyjs";
+import { StateTracker } from "./program";
+import {
+  type TransitionRes,
+  type WorkerState,
+  transitionStateArgsSchema,
+} from "./types";
+import { INITIAL_STATE, MERKLE_TREE_HEIGHT, MerkleWitness20 } from "./utils";
+
+const state: WorkerState = {
+  tree: new MerkleTree(MERKLE_TREE_HEIGHT),
+  transitionIndex: 0n, // should increment on state transition and is used to index the tree
+  updateQueue: [],
+  executingUpdate: false,
+};
+
+setInterval(() => {
+  if (
+    !state.latestProof ||
+    state.executingUpdate ||
+    !state.updateQueue.length
+  ) {
+    return;
+  }
+
+  const nextUpdate = state.updateQueue.shift();
+  if (!nextUpdate) return;
+
+  state.executingUpdate = true;
+
+  nextUpdate(state.latestProof)
+    .then((proof) => {
+      state.latestProof = proof;
+
+      const message: ZkappWorkerReponse = {
+        resType: "proof-update",
+        id: 0,
+        data: proof.toJSON(),
+      };
+      postMessage(message);
+
+      state.executingUpdate = false;
+    })
+    .catch((err) => {
+      console.warn("Update queue error:", err);
+      state.executingUpdate = false;
+    });
+}, 3000);
+
+const generateStateUpdate =
+  (newState: string, leafIndex: bigint) =>
+  async (prevProof: Proof<Field, void>) => {
+    const newStateCircuit = CircuitString.fromString(newState);
+
+    state.tree.setLeaf(leafIndex, Poseidon.hash(newStateCircuit.toFields()));
```

Comment:
Why do we need to convert to Field before hashing?
Is that the required input type for Poseidon here?